routing/router.go

```diff
@@ -3254,50 +3259,102 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
 			}
 		}
 
-		// Get an edge for the specific amount that we want to forward.
-		edge := edgeUnifier.getEdge(runningAmt, bandwidthHints, 0)
-		if edge == nil {
-			log.Errorf("Cannot find policy with amt=%v for hop %v",
-				runningAmt, i)
+		if i == len(unifiers)-1 {
+			// Get an edge for the specific amount that we want to
+			// forward.
+			edge = edgeUnifier.getEdge(
+				runningAmt, bandwidthHints, 0,
+			)
+			if edge == nil {
+				log.Errorf("Cannot find policy with amt=%v "+
+					"for hop %v", runningAmt, i)
 
-			return nil, 0, ErrNoChannel{
-				position: i,
+				return nil, 0, ErrNoChannel{
+					position: i,
+				}
```

Comment:
Split this case out of the loop, slice the edges down and run a single-branch loop for the remaining ones. We don't actually need to partition stuff since in both the outbound fee and inbound fee case there is only one exception and that exception is at one of the boundaries. This block will only ever be executed once, and it will be executed at the beginning.

Reply:
I chose it that way because `newRoute` has the same structure to have an easy comparison with it as the reference implementation. I'm aware that there is some code repetition, Joost tried to solve this as well in previous draft PRs, but generally I think that the current approach decouples the concerns of finding unified edges, which is not included in `newRoute`. The minimal amount search is a bit akin to do a first pathfinding stage that ends up with calling `newRoute` (this is how it's done in `findPath` as well) and kind of preexisting. Happy to change this to reduce the loop complexity if wanted

Reply:
Thanks for spelling that out, done!